test/object-store/sync/flx_sync.cpp

```diff
@@ -4995,6 +4967,464 @@ TEST_CASE("flx: nested collections in mixed", "[sync][flx][baas]") {
     CHECK(nested_list.get_any(1) == "foo");
 }
 
+const Schema g_person_schema{{"Person",
+                              {{"_id", PropertyType::ObjectId, Property::IsPrimary{true}},
+                               {"role", PropertyType::String},
+                               {"firstName", PropertyType::String},
+                               {"lastName", PropertyType::String}}}};
+
+auto fill_person_schema = [](SharedRealm realm, std::string role, size_t count) {
+    CppContext c(realm);
+    for (size_t i = 0; i < count; ++i) {
+        // Recreate the context (transaction) for every 100 entries
+        if (count % 100 == 0) {
+            c = CppContext(realm);
+        }
+        auto obj = Object::create(c, realm, "Person",
+                                  std::any(AnyDict{
+                                      {"_id", ObjectId::gen()},
+                                      {"role", role},
+                                      {"firstName", util::format("%1-%2", role, i)},
+                                      {"lastName", util::format("last-name-%1", i)},
+                                  }));
+    }
+};
+
+TEST_CASE("flx: role change bootstrap", "[sync][flx][baas][role_change][bootstrap]") {
+    enum BootstrapMode { NoReconnect, None, SingleMessage, SingleMessageMulti, MultiMessage, Any };
+
+    struct TestParams {
+        size_t num_emps = 500;
+        size_t num_mgrs = 10;
+        size_t num_dirs = 5;
+        std::optional<size_t> num_objects = std::nullopt;
+        std::optional<size_t> sleep_millis = std::nullopt;
+    };
+
+    struct ExpectedResults {
+        BootstrapMode bootstrap;
+        size_t emps;
+        size_t mgrs;
+        size_t dirs;
+    };
+
+    enum TestState { not_ready, start, reconnect_received, downloading, downloaded, complete };
+    TestingStateMachine<TestState> state_machina(TestState::not_ready);
+    std::mutex callback_mutex;
```

Comment:
seems like we always modify/read this variables around advancing/waiting on the state machine. do we need a separate mutex to protect them or is it sufficient to use waiting for the state machine to reach the desired state to fence these variables?